src/expr.rs

```diff
@@ -298,9 +308,57 @@ macro_rules! numeric (
 	($i:expr, $f:expr ) => (map_opt!($i,call!($f),EvalResult::as_numeric));
 );
 
+enum TypeCast {
+	SignedInt(usize),
+	UnsignedInt(usize),
+}
+
+impl EvalResult {
+	fn cast(self, ctype: TypeCast) -> Self {
+		use self::TypeCast::*;
+		let round = |bits| 2i64.pow((bits) as u32);
+		let unsigned_mask = |i, b| (i+round(b*8)) % round(b*8);
+		let signed_mask = |i, b| {
+			let rounded = unsigned_mask(i, b);
+			let half = round(8*b-1);
+			if rounded >= half {rounded - half * 2} else { rounded }
+		};
+		match self {
+			EvalResult::Float(f) => match ctype {
+				SignedInt(b) => EvalResult::Int(Wrapping(signed_mask(f as i64, b))),
+				UnsignedInt(b) => EvalResult::Int(Wrapping(unsigned_mask(f as i64, b))),
+			},
+			EvalResult::Int(Wrapping(i)) => match ctype {
+				SignedInt(b) => EvalResult::Int(Wrapping(signed_mask(i, b))),
+				UnsignedInt(b) => EvalResult::Int(Wrapping(unsigned_mask(i, b))),
+			},
+			_ => self
+		}
+	}
+}
+
+
 impl<'a> PRef<'a> {
+	method!(cast<PRef<'a>,&[Token],TypeCast,::Error>, mut self,
+		delimited!(p!("("), alt!(
+			do_parse!(k!("unsigned") >> k!("int") >> (TypeCast::UnsignedInt(size_of::<c_uint>()))) |
+			do_parse!(k!("unsigned") >> k!("short") >> (TypeCast::UnsignedInt(size_of::<c_ushort>()))) |
+			do_parse!(k!("unsigned") >> k!("char") >> (TypeCast::UnsignedInt(size_of::<c_uchar>()))) |
+			do_parse!(opt!(k!("signed")) >> k!("int") >> (TypeCast::SignedInt(size_of::<c_int>()))) |
+			do_parse!(opt!(k!("signed")) >> k!("short") >> (TypeCast::SignedInt(size_of::<c_short>()))) |
+			do_parse!(opt!(k!("signed")) >> k!("char") >> (TypeCast::SignedInt(size_of::<c_char>()))) |
+			do_parse!(i!("uint8_t") >> (TypeCast::UnsignedInt(1))) |
+			do_parse!(i!("uint16_t") >> (TypeCast::UnsignedInt(2))) |
+			do_parse!(i!("uint32_t") >> (TypeCast::UnsignedInt(4))) |
+			do_parse!(i!("int8_t") >> (TypeCast::SignedInt(1))) |
+			do_parse!(i!("int16_t") >> (TypeCast::SignedInt(2))) |
+			do_parse!(i!("int32_t") >> (TypeCast::SignedInt(4)))
```

Comment:
What about `long`, `long long`, and 64-bit types? In general, it may be useful to handle unknown types here in some graceful manner.

Reply:
Any integer type longer than 32 bits(size_t, uint64_t, long) is not supported since EvalResult::Int can not hold it. We need change the API to support 64-bit types.